src/posix/platform/backtrace.cpp

```diff
@@ -0,0 +1,170 @@
+#include <string.h>
+#include <unistd.h>
+
+#include "common/code_utils.hpp"
+#include "common/logging.hpp"
+
+#if OPENTHREAD_POSIX_CONFIG_BACKTRACE_ENABLE
+
+#if defined(__linux__) && !OPENTHREAD_ENABLE_ANDROID_NDK
+#if OPENTHREAD_POSIX_CONFIG_ANDROID_ENABLE
+#include <log/log.h>
+#include <utils/CallStack.h>
+#include <utils/Printer.h>
+
+class LogPrinter : public android::Printer
+{
+public:
+    virtual void printLine(const char *string) { otLogCritPlat("%s", string); }
+};
+
+static void dumpStack(void)
+{
+    LogPrinter         printer;
+    android::CallStack callstack;
+
+    callstack.update();
+    callstack.print(printer);
+}
+#else // OPENTHREAD_POSIX_CONFIG_ANDROID_ENABLE
+#include <cxxabi.h>
+#include <execinfo.h>
+
+static void demangleSymbol(int aIndex, const char *aSymbol)
+{
+    constexpr uint16_t kMaxNameSize = 256;
+    int                status;
+    char               program[kMaxNameSize + 1];
+    char               mangledName[kMaxNameSize + 1];
+    char *             demangledName = nullptr;
+    char *             functionName  = nullptr;
+    unsigned int       offset        = 0;
+    unsigned int       address       = 0;
+
+    // Try to demangle a C++ name
+    if (sscanf(aSymbol, "%256[^(]%*[^_]%256[^)+]%*[^0x]%x%*[^0x]%x", program, mangledName, &offset, &address) == 4)
+    {
+        demangledName = abi::__cxa_demangle(mangledName, nullptr, nullptr, &status);
+        functionName  = demangledName;
+    }
+
+    VerifyOrExit(demangledName == nullptr);
+
+    // If failed to demangle a C++ name, try to get a regular C symbol
+    if (sscanf(aSymbol, "%256[^(](%256[^)+]%*[^0x]%x%*[^0x]%x", program, mangledName, &offset, &address) == 4)
+    {
+        functionName = mangledName;
+    }
+
+exit:
+    if (functionName != nullptr)
+    {
+        otLogCritPlat("#%2d: %s %s+0x%x [0x%x]\n", aIndex, program, functionName, offset, address);
+    }
+    else
+    {
+        otLogCritPlat("#%2d: %s\n", aIndex, aSymbol);
+    }
+
+    if (demangledName != nullptr)
+    {
+        free(demangledName);
+    }
+}
+
+static void dumpStack(void)
+{
+    constexpr uint8_t kMaxDepth = 50;
+    void *            stack[kMaxDepth];
+    char **           symbols;
+    int               depth;
+
+    depth   = backtrace(stack, kMaxDepth);
+    symbols = backtrace_symbols(stack, depth);
+    VerifyOrExit(symbols != nullptr);
+
+    for (int i = 0; i < depth; i++)
+    {
+        demangleSymbol(i, symbols[i]);
+    }
+
+    free(symbols);
+
+exit:
+    return;
+}
+#endif // OPENTHREAD_POSIX_CONFIG_ANDROID_ENABLE
+
+static void signalCritical(int sig, siginfo_t *info, void *ucontext)
+{
+    OT_UNUSED_VARIABLE(ucontext);
+    OT_UNUSED_VARIABLE(info);
+
+    otLogCritPlat("------------------ BEGINNING OF CRASH -------------");
+    otLogCritPlat("*** FATAL ERROR: Caught signal: %d (%s)", sig, strsignal(sig));
+
+    dumpStack();
+
+    otLogCritPlat("------------------ END OF CRASH ------------------");
+    exit(EXIT_FAILURE);
```

Comment:
Should we reset the signal handler to default and re-raise the original signal?

Reply:
Yes, it could avoid the OS from linking the default signal process handler. For example, when compiling the otbr-agent in Android, the Android will link the function [debuggerd_init()](https://android.googlesource.com/platform/bionic/+/06abcef/linker/debugger.cpp#301) and call it before calling the function `main()`.  The function `debuggerd_init()` sets the signal handler to catch the signal. So reset the signal handler to default and re-raise the original signal could make the OS also can catch signals.

I will submit a PR to process this case later.

Reply:
I have pushed a commit https://github.com/openthread/openthread/pull/8934 to handle this case.